[v7]: Better upgrade guide (#5617)
pulumi/pulumi-aws: docs/version-7-upgrade.md

- `endpoints.worklink` – removed due to the removal of Amazon Worklink support.

## Enhanced Region Support

Version 7.0.0 adds `region` to most resources making it significantly easier to manage infrastructure across AWS Regions without requiring multiple provider configurations. See [Enhanced Region Support](./enhanced-region-support.md).

## Provider role chaining

v7 adds support for IAM role chaining by changing the `assumeRole` property to a
list that accepts multiple `assumeRole` arguments.

This feature was added to the upstream provider in the previous major version,
but since it required a breaking change in the Pulumi AWS SDK the change was
backed out.

If you are providing the `assumeRole` property on the `Provider`, you will need
to update your code. See the below example.

This (v6)

```ts
import * as aws from '@pulumi/aws';

const provider = new aws.Provider("provider", {
    assumeRole: {roleArn: baseRole.arn},
});

```

Becomes this (v7)

```ts
import * as aws from '@pulumi/aws';

const provider = new aws.Provider("provider", {
    assumeRoles: [{roleArn: baseRole.arn}],
});
```

## Bucket/BucketV2 Changes

> As part of this major version upgrade we are removing the old `Bucket` resource

Reviewer: Should we call it S3 Bucket instead? I know it's kind of assumed but perhaps not 100% obvious to all users. The heading should also be "S3 Bucket/BucketV2 Changes"

- `aws.s3.BucketServerSideEncryptionConfigurationV2` => `aws.s3.BucketServerSideEncryptionConfiguration`
- `aws.s3.BucketVersioningV2` => `aws.s3.BucketVersioning`
- `aws.s3.BucketWebsiteConfigurationV2` => `aws.s3.BucketWebsiteConfiguration`

### MaxItemsOne Changes

The following input properties have changed from a list of properties to a
single property.

- `loggings` => `logging`
- `replicationConfigurations` => `replicationConfiguration`
  - `rules`
    - `destinations` => `destination`
      - `replicationTimes` => `replicationTime`
      - `accessControlTranslations` => `accessControlTranslation`
    - `filters` => `filter`
    - `sourceSelectionCriterias` => `sourceSelectionCriteria`
      - `sseKmsEncryptedObjects` => `sseKmsEncryptedObjects` (list to object)
- `serverSideEncryptionConfigurations` => `serverSideEncryptionConfiguration`
  - `rules` => `rule`
    - `applyServerSideEncryptionByDefaults` => `applyServerSideEncryptionByDefault`
- `versionings` => `versioning`
- `website` => `websites`
- `lifecycleRules`
  - `noncurrentVersionExpirations` => `noncurrentVersionExpiration`
  - `expirations` => `expiration`

### Migration

> In order to perform this migration first update your code to use the new resource or property names and then run pulumi with the `--refresh` and `--run-program` arguments. Make sure you have [installed](https://www.pulumi.com/docs/iac/download-install/) the latest version of the Pulumi CLI.

Reviewer: This might be a bit easier to understand as a step-by-step list:

1. Change program from `Bucketv2` to `Bucket1
2. Change resource names to match the new ones
3. Run `pulumi up --refresh --run-program`
4. Ensure there is no diff in the update. If there is, go back to step 1

            "s3:GetObjectVersionAcl",
            "s3:GetObjectVersionTagging"
          ],
          Effect: "Allow",
          Resource: [
            `${migrationBucket.arn}/*`
          ]
        },
        {
          Action: [
            "s3:ReplicateObject",
            "s3:ReplicateDelete",
            "s3:ReplicateTags"
          ],
          Effect: "Allow",
          Resource: `${destinationBucket.arn}/*`
        }
      ]
    }),
}, { provider: awsProvider });

// Replication Role Policy Attachment
const replicationRolePolicyAttachment = new aws.iam.RolePolicyAttachment("replication", {
    role: replication.name,
    policyArn: replicationPolicy.arn,
}, { provider: awsProvider });
```

## Tags Changes

This release some changes to the way tags work in the Pulumi AWS Provider. In v6
of the Pulumi AWS Provider we customized the tagging behavior of the provider
through Pulumi level patches to the upstream Terraform provider along with other
Pulumi level customizations. This resulted in tagging behavior that diverged
from the upstream Terraform provider and has been difficult to maintain as the
upstream Terraform provider has made changes to tagging.

> In v7 we are removing the Pulumi level customizations and going back to relying

Reviewer: What is the implication for users here? They probably don't care about the implementation that much.

on the upstream provider's tagging behavior.

## Patch Related Changes

> We have created several Pulumi level patches to the upstream Terraform provider

Reviewer: ```suggestion
We previously had Pulumi level patches to the upstream Terraform provider
```

makes it a bit more obvious that it's not a new thing
